### PR TITLE
interpret/cast: make more matches on FloatTy properly exhaustive

### PR DESCRIPTION
**Pull request by RalfJung**
Actually implementing these is pretty trivial (at least once all the scalar methods are added, which happens in https://github.com/rust-lang/rust/pull/121926), but I'm staying consistent with the other f16/f128 PRs. Also adding adding all the tests to Miri would be quite a lot of work.

There's probably some way to reduce the code duplication here with more use of generics... but that's a future refactor.^^

r? @tgross35

### CODE REVIEW

**Comment by RalfJung** [comment]
r? compiler

**Review by tgross35 — APPROVED** [review]
Can't r but this looks great, thanks for making my work easier :)

**Review by compiler-errors — APPROVED** [review]
thanks ralf, i love exhaustive matches

File: compiler/rustc_const_eval/src/interpret/cast.rs [hunk 1]
```diff
@@ -285,15 +287,26 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
+            Float(fty) if signed => {
+                let v = v as i128;
+                match fty {
+                    FloatTy::F16 => unimplemented!("f16_f128"),
+                    FloatTy::F32 => Scalar::from_f32(Single::from_i128(v).value),
+                    FloatTy::F64 => Scalar::from_f64(Double::from_i128(v).value),
+                    FloatTy::F128 => unimplemented!("f16_f128"),
+                }
             }
+            // unsiged int -> float
```
Comment (lqd) [inline compiler/rustc_const_eval/src/interpret/cast.rs:300]:
```suggestion
            // unsigned int -> float
```

**Comment by tgross35** [comment]
> Also adding adding all the tests to Miri would be quite a lot of work

Talk about great timing. Getting Miri tests to work with stage0 was a pain point for the proof of concept at https://github.com/rust-lang/rust/pull/114607/files#diff-24bd7b5e2ca32dd28044509d08e68ec5be1c382735128742baec9383acbd6fea. But with a beta branch in only 11 days, that won't be a concern as long as https://github.com/rust-lang/rust/pull/121926 lands before then 🎉

**Comment by RalfJung** [comment]
> > Also adding adding all the tests to Miri would be quite a lot of work
> 
> Talk about excellent timing. Getting Miri tests to work with stage0 was a pain point for the proof of concept at https://github.com/rust-lang/rust/pull/114607/files#diff-24bd7b5e2ca32dd28044509d08e68ec5be1c382735128742baec9383acbd6fea. But with a beta branch in only 11 days, that won't be a concern as long as #121926 and some libs support land before then 🎉

That's only an issue for operations that need host floats. Most operations use soft floats which you can implement without any support in the bootstrap compiler.

The tests are a lot of work just because there's so many float operations.^^

**Pull request merged**